### PR TITLE
[Fixed SD-275] Investment math

### PR DESCRIPTION
**Pull request by mightymatth**

### CODE REVIEW

File: src/app/invest/invest.component.ts [hunk 1]
```diff
@@ -19,31 +21,71 @@ import {RouterService} from '../shared/services/router.service'
...
+
+        const campaignTokensBalance = Number(utils.formatEther(campaign.totalTokensBalance))
+        const campaignTokensSold = Number(utils.formatEther(campaign.totalTokensSold))
+        const campaignTokenPrice = TokenPrice.parse(campaign.tokenPrice.toNumber())
+        const campaignTokensAvailable = campaignTokensBalance - campaignTokensSold
+        const campaignAmountAvailable = campaignTokensAvailable * campaignTokenPrice
+
+        const walletBalance = Number(utils.formatEther(balance))
+
+        let min = alreadyInvested > 0 ? 0 : campaignMin
```
Comment (githinho) [inline src/app/invest/invest.component.ts:78]:
```suggestion
        const min = alreadyInvested > 0 ? 0 : campaignMin
```

File: src/app/invest/invest.component.ts [hunk 2]
```diff
@@ -19,31 +21,71 @@ import {RouterService} from '../shared/services/router.service'
...
+        let max = campaignAmountAvailable < campaignMax ? campaignAmountAvailable : campaignMax
+
+        // ensure floor on 2 decimal points
+        max = Math.round(Math.floor(max * 100)) / 100
```
Comment (githinho) [inline src/app/invest/invest.component.ts:82]: Why use `round` after `floor`? Looks to me that `floor` produces a round number: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/floor

File: src/app/campaigns/campaign-detail/campaign-detail.component.html [hunk 3]
```diff
@@ -108,8 +108,18 @@ <h1 class="my-2 text-2xl uppercase">
...
         <div>
-          Total investors count:
-          {{ campaign.totalInvestorsCount | formatUnit:'toNumber' | number }}
+          Total funds raised:
+          {{ campaign.totalFundsRaised | formatUnit:'bignumToWei' | currencyDefault }}
+        </div>
+
+        <div>
+          Total tokens sold:
+          {{ campaign.totalTokensSold| formatUnit:'bignumToWei' | number }}
```
Comment (mightymatth) [inline src/app/campaigns/campaign-detail/campaign-detail.component.html:117]: spacing

File: src/app/invest/invest.component.ts [hunk 4]
```diff
@@ -124,10 +159,10 @@ export class InvestComponent {
 
   invest() {
     return combineLatest([
-      this.campaign$.pipe(filter(res => !!res.value)),
+      this.campaign$,
     ]).pipe(take(1),
       switchMap(([campaign]) => this.campaignService.invest(
-        campaign.value!.contractAddress,
+        campaign!.contractAddress,
```
Comment (githinho) [inline src/app/invest/invest.component.ts:165]: Is `!` really needed? You didn't use it in function `approveAmount`.

**Review by domagojlatecki — APPROVED** [review]

File: src/app/invest/invest.component.ts (hunk 2 above)
Comment (mightymatth) [reply src/app/invest/invest.component.ts:82]: Sure!
```
Math.floor(max * 100) / 100
```

File: src/app/invest/invest.component.ts [hunk 5]
```diff
@@ -64,25 +63,25 @@ export class InvestComponent {
...
 
         const min = alreadyInvested > 0 ? 0 : campaignMin
-        let max = campaignAmountAvailable < campaignMax ? campaignAmountAvailable : campaignMax
 
-        // ensure floor on 2 decimal points
-        max = Math.round(Math.floor(max * 100)) / 100
+        let userInvestGap = this.floorDecimals(campaignMax - alreadyInvested)
+        let max = this.floorDecimals(Math.min(userInvestGap, campaignStats.valueToInvest))
+
+        userInvestGap = Math.floor(userInvestGap * 100) / 100
```
Comment (githinho) [inline src/app/invest/invest.component.ts:78]: Why not use the function `floorDecimals`?

```suggestion
        userInvestGap = floorDecimals(userInvestGap)
```

File: src/app/shared/components/funding-progress/funding-progress.component.ts [hunk 6]
```diff
@@ -1,22 +1,44 @@
...
+      stats.softCap / stats.valueTotal : 0
+
+    this.progressData = {
+      raised: stats.valueInvested,
+      raisedPercentage: raisedPercentage,
+      total: stats.valueTotal,
+      softCap: stats.softCap,
+      softCapPercentage: softCapPercentage,
+    }
+    console.log(this.progressData)
```
Comment (githinho) [inline src/app/shared/components/funding-progress/funding-progress.component.ts:34]: Do we need to log this?

**Review by githinho — APPROVED** [review]

**Pull request merged**